pkg/healthchecker/health_checker_windows.go

```diff
@@ -0,0 +1,190 @@
+		}
+		out = strings.ReplaceAll(out, ",", "")
+		t, err := time.Parse(types.UptimeTimeLayout, out)
+		if err != nil {
+			return time.Duration(0), err
+		}
+		return time.Since(t), nil
+	}
+}
+
+// getRepairFunc returns the repair function based on the component.
+func getRepairFunc(hco *options.WindowsHealthCheckerOptions) func() {
+	// Restart-Service will stop and attempt to start the service
+	return func() {
+		execCommand("powershell.exe", "Restart-Service", hco.Service)
+	}
+}
+
+// getHealthCheckFunc returns the health check function based on the component.
+func getHealthCheckFunc(hco *options.WindowsHealthCheckerOptions) func() (bool, error) {
+	switch hco.Component {
+	case types.KubeletComponent:
+		return func() (bool, error) {
+			httpClient := http.Client{Timeout: hco.HealthCheckTimeout}
+			response, err := httpClient.Get(types.KubeletHealthCheckEndpoint)
+			if err != nil || response.StatusCode != http.StatusOK {
+				return false, nil
+			}
+			return true, nil
+		}
+	case types.DockerComponent:
+		return func() (bool, error) {
+			if _, err := execCommand("powershell.exe", "docker", "ps"); err != nil {
+				return false, nil
+			}
+			return true, nil
+		}
+	case types.CRIComponent:
+		return func() (bool, error) {
+			if _, err := execCommand("powershell.exe", hco.CriCtlPath, "--runtime-endpoint="+hco.CriSocketPath, "--image-endpoint="+hco.CriSocketPath, "pods"); err != nil {
+				return false, nil
+			}
+			return true, nil
+		}
+	}
+	return nil
+}
+
+// execCommand executes the bash command and returns the (output, error) from command, error if timeout occurs.
+func execCommand(command string, args ...string) (string, error) {
+	cmd := util.Exec(command, args...)
+	out, err := cmd.Output()
+	if err != nil {
+		glog.Infof("command %v failed: %v, %v\n", cmd, err, out)
+		return "", err
+	}
+	return strings.TrimSuffix(string(out), "\r\n"), nil
+}
+
+// checkForPattern returns (true, nil) if logPattern occurs atleast logCountThreshold number of times since last
+// service restart. (false, nil) otherwise.
+func checkForPattern(service, logStartTime, logPattern string, logCountThreshold int) (bool, error) {
+	countPatternLogCmd := "@(Get-WinEvent -Logname System | Where-Object {($_.TimeCreated -ge ([datetime]::ParseExact('" + logStartTime +
+		"','" + types.LogParsingTimeFormat + "', $null))) -and ($_.Message -Match '" + logPattern + "')}).count"
+
+	out, err := execCommand("powershell.exe", countPatternLogCmd)
+	if err != nil {
+		return true, err
+	}
+	occurrences, err := strconv.Atoi(out)
+	if err != nil {
+		return true, err
+	}
+	if occurrences >= logCountThreshold {
+		glog.Infof("%s failed log pattern check, %s occurrences: %v", service, logPattern, occurrences)
+		return false, nil
+	}
+	return true, nil
+}
+
+// CheckHealth checks for the health of the component and tries to repair if enabled.
+// Returns true if healthy, false otherwise.
+func (hc *windowsHealthChecker) CheckHealth() (bool, error) {
```

Comment:
You can split a struct definition between 2 files. Also this function is exactly the same in both OS with the exception of the service name field which should be renamed.

Reply:
I refactored to share most of the items, and factored out only specific linux/windows things.